src/gt4py/next/new_type_system/types.py

```diff
@@ -0,0 +1,319 @@
+from . import traits
+import dataclasses
+from typing import Optional, Any, Sequence
+
+from .traits import FunctionArgument
+
+
+class Type:
+    ...
+
+
+@dataclasses.dataclass
+class IntegerType(
```

Comment:
These `IntegerType` and `FloatType` definitions make sense to me. However, the concepts here deviate a bit from the most common conventions in array/tensor libraries used in the community. Ideally, we could represent arithmetic types in a way that is compatible with both dtype-like descriptors used by array libraries at runtime for data exchange and allocation, and internally in our description of the ASTs. A first step into this direction are the DType definitions on `gt4py._core.definitions` based on DLpack (https://dmlc.github.io/dlpack/latest/) and the Python Array API standard (https://data-apis.org/array-api/latest/API_specification/type_promotion.html#type-promotion). The current implementation there still needs to be polished and extended so I think it would make sense to try to unify those with the definitions here, instead of having two similar but incompatible representations for the same concept.

Actually, I think the core part of the `new_type_system` could be moved into `gt4py._core` to make things simpler.

Reply:
I don't really think it's a good idea to include these tasks in this PR. First, the goal of this PR was to find a cleaner representation of AST types, and it delivers that in a self-contained package. Turning it into an open-ended task would just delay it and reduce quality. Second, if the DType definitions and their use cases and usage patterns are not clear, it's difficult to decide if merging makes sense at all, and if it does, how it should be done. And, third, this type system is only for `gt4py.next`, so unless it's backported to the cartesian version, I don't see the benefits of lifting it out of `next`.